### ISSUE TITLE
torch.compile hang/crashes with worker_start_method=spawn

### ISSUE THREAD

**Issue by ppwwyyxx** [pytorch/pytorch#126311]
### 🐛 Describe the bug

The following code:
```python
import torch
import torch._inductor.config

torch._inductor.config.worker_start_method = 'spawn'
torch._inductor.config.compile_threads = 2
from torch._inductor.codecache import AsyncCompile


if __name__ == '__main__':
    model = torch.nn.Linear(10,10).cuda()
    model = torch.compile(model)
    inp = torch.rand(10, 10).cuda()
    model(inp).sum().backward()
    print("Done")
```
crashes with:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/spawn.py", line 237, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/spawn.py", line 288, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/runpy.py", line 269, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/runpy.py", line 96, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/XXX/a.py", line 6, in <module>
    from torch._inductor.codecache import AsyncCompile
  File "/home/XXX/.pyenv/versions/torch22/lib/python3.10/site-packages/torch/_inductor/codecache.py", line 2492, in <module>
    AsyncCompile.warm_pool()
  File "/home/XXX/.pyenv/versions/torch22/lib/python3.10/site-packages/torch/_inductor/codecache.py", line 2422, in warm_pool
    pool._adjust_process_count()
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/concurrent/futures/process.py", line 683, in _adjust_process_count
    p.start()
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/spawn.py", line 155, in get_preparation_data
    _check_not_importing_main()
  File "/home/XXX/.pyenv/versions/3.10.0/lib/python3.10/multiprocessing/spawn.py", line 135, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...
```

The following code:
```
import torch
import torch._inductor.config
torch._inductor.config.worker_start_method = 'spawn'
torch._inductor.config.compile_threads = 2

if __name__ == '__main__':
    model = torch.nn.Linear(10,10).cuda()
    model = torch.compile(model)
    inp = torch.rand(10, 10).cuda()
    model(inp).sum().backward()
    print("Done")
```
hangs.

Changing to `fork`, both code succeeds.

### Versions

```
PyTorch version: 2.2.2+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.3) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.10.0 (default, Dec 18 2023, 03:34:21) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.250-2-velinux1u1-amd64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
```

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

**Comment by xmfan** [comment]
Repro'd on 2.3 stable,

Looks like we've added spawn support last year: https://github.com/pytorch/pytorch/pull/108850, but it might be broken now

cc @oulgen @masnesral

**Comment by masnesral** [comment]
I think spawn is probably not a good option anyway since it re-executes the entry script and we can't enforce that it checks for '__main__'.  I'm hoping to provide a "safe" parallel compile alternative in the next few weeks that will popen a sub-process instead of fork. After that, I'd vote to remove the spawn support altogether.

@ppwwyyxx is there a specific reason you need "spawn" (other than the existing fork-based impl can be problematic)?

**Comment by ppwwyyxx** [comment]
The reason is just that fork is problematic as expected. Having a subprocess option would be great.

**Comment by masnesral** [comment]
@ppwwyyxx we now have "subprocess" as on option instead of "spawn" or "fork" (in fact it's the default in recent revs). Do you want to give it a try?

**Comment by arogozhnikov** [comment]
@masnesral 

I use 2.3.1 and spawn seems to be indeed broken.

Tried  `config.worker_start_method = 'subprocess'` , result is `ValueError: cannot find context for 'subprocess'.` (still in nightly?)

**Comment by arogozhnikov** [comment]
(forkserver is broken too)

**Comment by masnesral** [comment]
@arogozhnikov yeah looks like the fix is not in 2.3.1 and you'd have to get a recent nightly build.

**Comment by Edenzzzz** [comment]
Hitting this with torch 2.1.1 and python 3.9
